forest/shared/src/snapshot.rs

```
@@ -0,0 +1,569 @@
+/// return [`Ok(Vec::new())`] if `snapshot_directory` does not exist.
+///
+/// Users can freely delete the path to the blob - corresponding metadata will
+/// be cleaned up in the next call to [list], but this should be regarded as an
+/// implementation detail.
+///
+/// See [module documentation](mod@self) for more.
+///
+/// Note this function makes blocking syscalls, and should not be called
+/// from an async context. Use [`tokio::task::spawn_blocking`] if needed.
+pub fn list(snapshot_directory: &Path) -> anyhow::Result<Vec<(PathBuf, SnapshotMetadata)>> {
+    if !snapshot_directory.exists() {
+        return Ok(Vec::new());
+    }
+    // Get all the file paths
+    let mut paths = walkdir::WalkDir::new(snapshot_directory)
+        .min_depth(1)
+        .max_depth(1)
+        .into_iter()
+        .filter_ok(|entry| entry.path().is_file())
+        .map_ok(|entry| match entry.path().to_str() {
+            // prefix operation on strings are easier, so convert to those
+            Some(s) => Some(String::from(s)),
+            None => {
+                warn!(path = %entry.path().display(), "ignored non-utf8 file in snapshot directory");
+                None
+            }
+        })
+        .flatten_ok()
+        .collect::<Result<BTreeSet<_>, _>>()
+        .context("couldn't enumerate paths in snapshot directory")?;
+
+    // Sort them into pairs
+    let mut blobs_and_metadata = Vec::new();
+    while let Some(path) = paths.pop_first() {
+        match path.strip_suffix(METADATA_FILE_SUFFIX) {
+            Some(blob_path) => {
+                // we've popped the metadata file, try and pop the blob path
+                let blob_was_present = paths.remove(blob_path);
+                match blob_was_present {
+                    false => {
+                        warn!(%path, "deleting metadata without corresponding blob");
+                        let _ = std::fs::remove_file(path);
```

why do we need `let _` ?

---

Thanks, I didn't know of this idiom. Why are we ignoring the error? Shouldn't we at least log the error? Otherwise, potential issues may fly under our radar.

---

Closing this as this will all be ripped out anyway